joomla/joomla-cms
# JRoute between clients (site -> admin | admin -> site)

Pull Request for Improvement.
#### Summary of Changes

This PR is for allowing JRoute to generate URL's from admin to site and vice-versa.

This is a current problem especially if you are trying to make links from the backend to the frontend.

So, with this PR, there is no need anymore to make special code just to get the frontend link from the backend. With this PR we just need to use JRoute, like all other links.

The same will work to links from frontend to backend.
After this PR, no need to hardcode frontend to admin links.
#### Testing Instructions
##### Test 1: Links from admin to site
- Use latest staging with sample data from at least two languages (en-GB, fr-FR)
- Add to isis index.php file...

``` php
$app = JFactory::getApplication();
$app->enqueueMessage('1. Admin to Admin: ' . JRoute::_('index.php'), 'notice');
$app->enqueueMessage('2. Admin to Admin: ' . JRoute::_('index.php?option=com_content'), 'notice');
$app->enqueueMessage('3. Admin to Site: ' . JRoute::_('index.php?option=com_content', true, null, 'site'), 'notice');
$app->enqueueMessage('4. Admin to Site (MutilLanguage 1): ' . JRoute::_('index.php?option=com_content&lang=fr', true, null, 'site'), 'notice');
$app->enqueueMessage('5. Admin to Site (MutilLanguage 2): ' . JRoute::_('index.php?option=com_users&lang=en', true, null, 'site'), 'notice');
```
- Open any admin page. Check the URL in resulting notice message are correct.

Note: You should test also other URLs (articles, menu items, etc).

Note 2: Try with the several variation of SEF/non-SEF,  with or without URL rewritting,  with or without multilanguage (i.e., with or without language filter plugin enabled).
##### Test 2: Links from site to admin

Add to protostar index.php file...

``` php
$app = JFactory::getApplication();
$app->enqueueMessage('1. Site to Site: ' . JRoute::_('index.php'), 'notice');
$app->enqueueMessage('2. Site to Site: ' . JRoute::_('index.php?option=com_content'), 'notice');
$app->enqueueMessage('3. Site to Admin: ' . JRoute::_('index.php?option=com_content', true, null, 'administrator'), 'notice');
```
- Open any frontend page. Check the URL in resulting notice message are correct.

Note: You should test also other URLs.

Note 2: Try with the several variation of SEF/non-SEF,  with or without URL rewritting,  with or without multilanguage (i.e., with or without language filter plugin enabled).
##### Test 3: Code review

Do a code review.
##### Test 4: Current URLs

Check if current URL generated by JRouter are exactly the same, i.e., no change in current URL.
Probably the best is to leave the patch applied for some time, while you do other things and check there are no problems with the current URLs.
#### Observations

Thanks to @mbabker, @ggppdk and @infograf768 for helping out with this one.

Mantainers, to make sure that are no unintended consequences, it would be good if this can be also reviewed from someone that understand the Router better than i do.
